psydac/api/tests/test_api_2d_compatible_spaces.py

```diff
@@ -235,34 +271,102 @@ def run_system_3_2d_dir(uex, f, alpha, ncells, degree):
 ###############################################################################
 #            SERIAL TESTS
 ###############################################################################
-def test_api_system_1_2d_dir_1():
+def test_poisson_mixed_form_2d_dir_1():
     from sympy import symbols
     x1, x2 = symbols('x1, x2')
 
     f0 =  -2*x1*(1-x1) -2*x2*(1-x2)
     u  = x1*(1-x1)*x2*(1-x2)
 
-    l2_error = run_system_1_2d_dir(f0,u, ncells=[2**3,2**3], degree=[2,2])
+    l2_error = run_poisson_mixed_form_2d_dir(f0, u, ncells=[2**3, 2**3], degree=[2, 2])
     assert l2_error-0.00030070020628128664<1e-13
 
 #------------------------------------------------------------------------------
-def test_api_system_2_2d_dir_1():
+@pytest.mark.parametrize('scipy', (True, False))
+def test_stokes_2d_dir_homogeneous(scipy):
 
-    from sympy import symbols
-    x1, x2 = symbols('x1, x2')
+    # ... Exact solution
+    domain = Square()
+    x, y   = domain.coordinates
  
-    f1 = -x1**2*(x1 - 1)**2*(24*x2 - 12) - 4*x2*(x1**2 + 4*x1*(x1 - 1) + (x1 - 1)**2)*(2*x2**2 - 3*x2 + 1) - 2*pi*cos(2*pi*x1)
-    f2 = 4*x1*(2*x1**2 - 3*x1 + 1)*(x2**2 + 4*x2*(x2 - 1) + (x2 - 1)**2) + x2**2*(24*x1 - 12)*(x2 - 1)**2 + 2*pi*cos(2*pi*x2)
-    u1 = x1**2*(-x1 + 1)**2*(4*x2**3 - 6*x2**2 + 2*x2)
-    u2 =-x2**2*(-x2 + 1)**2*(4*x1**3 - 6*x1**2 + 2*x1)
-    p  = sin(2*pi*x1) - sin(2*pi*x2)
-    
-    l2_error = run_system_2_2d_dir(f1, f2, u1, u2, ncells=[2**4,2**4], degree=[2,2])
-    assert l2_error-0.020113712082281063<1e-13
-    #TODO verify convergence rate
+    fx = -x**2*(x - 1)**2*(24*y - 12) - 4*y*(x**2 + 4*x*(x - 1) + (x - 1)**2)*(2*y**2 - 3*y + 1) - 2*pi*cos(2*pi*x)
+    fy = 4*x*(2*x**2 - 3*x + 1)*(y**2 + 4*y*(y - 1) + (y - 1)**2) + y**2*(24*x - 12)*(y - 1)**2 + 2*pi*cos(2*pi*y)
+    f  = Tuple(fx, fy)
+
+    ux = x**2*(-x + 1)**2*(4*y**3 - 6*y**2 + 2*y)
+    uy =-y**2*(-y + 1)**2*(4*x**3 - 6*x**2 + 2*x)
+    ue = Tuple(ux, uy)
+    pe = -sin(2*pi*x) + sin(2*pi*y)
+    # ...
+
+    # Verify that div(u) = 0
+    assert (ux.diff(x) + uy.diff(y)).simplify() == 0
+
+    # ... Check that exact solution is correct
+    from sympde.calculus import laplace, grad
+    from sympde.expr import TerminalExpr
+
+    kwargs = dict(dim=2, logical=True)
+    a = TerminalExpr(-laplace(ue), **kwargs)
+    b = TerminalExpr(    grad(pe), **kwargs)
+    c = TerminalExpr(   Matrix(f), **kwargs)
+    err = (a.T + b - c).simplify()
+
+    assert err[0] == 0
+    assert err[1] == 0
+    # ...
```

Comment:
this is nice !

Reply:
:sweat_smile: Yes, it's nice to check that the exact solution is correct!